File: leetcode/0383_ransom_note.cpp
// Copyright (c) 2018 Jacopo Notarstefano

#include <cassert>
#include <string>
#include <unordered_map>

using namespace std;

class Solution {
 public:
    bool canConstruct(string ransomNote, string magazine) {
        unordered_map<char, int> ransomNoteCounter;
        for (auto c : ransomNote) {
            ransomNoteCounter[c]++;
        }

        unordered_map<char, int> magazineCounter;
        for (auto c : magazine) {
            magazineCounter[c]++;
        }

        for (auto pair : ransomNoteCounter) {
            auto it = magazineCounter.find(pair.first);
            if (it == magazineCounter.end() || pair.second > it->second) {
                return false;
            }
        }

        return true;
    }
};

int main() {
    auto solution = Solution();

    assert(!solution.canConstruct("a", "b"));
    assert(!solution.canConstruct("aa", "ab"));
    assert(solution.canConstruct("aa", "aab"));
}

File: leetcode/0383_ransom_note.py
# -*- coding: utf-8 -*-

from collections import Counter


class Solution:
    def canConstruct(self, ransomNote, magazine):
        magazine_counter = Counter(magazine)
        ransomNote_counter = Counter(ransomNote)
        magazine_counter.subtract(ransomNote_counter)

        return all(el >= 0 for el in magazine_counter.values())


if __name__ == '__main__':
    solution = Solution()

    assert not solution.canConstruct('a', 'b')
    assert not solution.canConstruct('aa', 'ab')
    assert solution.canConstruct('aa', 'aab')
